File: elixir/math/modpow.exs
# Print modular powers of fixed triples, each squared and halved by repeated squaring.
# Run: elixir modpow.exs

defmodule ModPow do
  def compute(base, exponent, modulus), do: walk(rem(base, modulus), exponent, modulus, 1)

  defp walk(_base, 0, _modulus, result), do: result

  defp walk(base, exponent, modulus, result) do
    result = if rem(exponent, 2) == 1, do: rem(result * base, modulus), else: result
    walk(rem(base * base, modulus), div(exponent, 2), modulus, result)
  end
end

[{2, 1000, 1000003}, {3, 200, 50}, {5, 117, 19}, {10, 18, 9999991}]
|> Enum.each(fn {base, exponent, modulus} ->
  IO.puts("#{base} #{exponent} #{modulus} #{ModPow.compute(base, exponent, modulus)}")
end)

File: ocaml/math/modpow.ml
(* Print modular powers of fixed triples, each squared and halved by repeated squaring. *)
(* Run: ocaml modpow.ml *)

let cases = [ (2, 1000, 1000003); (3, 200, 50); (5, 117, 19); (10, 18, 9999991) ]

let modpow base exponent modulus =
  let rec walk base exponent result =
    if exponent = 0 then result
    else
      let result = if exponent mod 2 = 1 then result * base mod modulus else result in
      walk (base * base mod modulus) (exponent / 2) result
  in
  walk (base mod modulus) exponent 1

let () =
  List.iter
    (fun (base, exponent, modulus) ->
      Printf.printf "%d %d %d %d\n" base exponent modulus (modpow base exponent modulus))
    cases
